src/main/java/de/csbdresden/csbdeep/ui/GPUSettingsFrame.java

```diff
@@ -0,0 +1,182 @@
+		gpuChoiceBox.addItem("GPU");
+		gpuChoiceBox.addItem("CPU");
+		gpuChoiceBox.addActionListener(new AbstractAction() {
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				updateFilter();
+			}
+		});
+		cudaChoiceBox = new JComboBox<>();
+		cudaChoiceBox.addActionListener(new AbstractAction() {
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				updateFilter();
+			}
+		});
+		tfChoiceBox = new JComboBox<>();
+		tfChoiceBox.addActionListener(new AbstractAction() {
+			@Override
+			public void actionPerformed(ActionEvent e) {
+				updateFilter();
+			}
+		});
+		JPanel panel = new JPanel();
+		panel.setLayout(new MigLayout());
+		panel.add(new JLabel("Filter by.."));
+		panel.add(makeLabel("Mode: "));
+		panel.add(gpuChoiceBox);
+		panel.add(makeLabel("CUDA: "));
+		panel.add(cudaChoiceBox);
+		panel.add(makeLabel("TensorFlow: "));
+		panel.add(tfChoiceBox);
+		return panel;
+	}
+
+	private void updateFilter() {
+		installPanel.removeAll();
+		buttons.forEach(btn -> {
+			if(filter("", gpuChoiceBox, btn)) return;
+			if(filter("CUDA ", cudaChoiceBox, btn)) return;
+			if(filter("TF ", tfChoiceBox, btn)) return;
+			installPanel.add(btn);
+		});
+		installPanel.revalidate();
+		installPanel.repaint();
+	}
+
+	private boolean filter(String title, JComboBox<String> choiceBox, JRadioButton btn) {
+		if(choiceBox.getSelectedItem().toString().equals(NOFILTER)) return false;
+		return !btn.getText().contains(title + choiceBox.getSelectedItem().toString());
+	}
+
+	private JLabel makeLabel(String s) {
+		JLabel label = new JLabel(s);
+		label.setHorizontalAlignment(SwingConstants.RIGHT);
+		label.setHorizontalTextPosition(SwingConstants.RIGHT);
+		label.setBackground(Color.red);
+		return label;
+	}
+
+	private Component createInstallPanel() {
+		installPanel = new JPanel(new MigLayout("flowy"));
+		return new JScrollPane(installPanel);
+	}
+
+	private Component createActionPanel() {
+		JPanel panel = new JPanel();
+		return panel;
+	}
+
+	public void updateChoices(List<LibraryVersion> availableVersions) {
+		availableVersions.sort(Comparator.comparing(LibraryVersion::getOrderableTFVersion));
+		this.availableVersions = availableVersions;
+		updateCUDAChoices();
+		updateTFChoices();
+		versionGroup = new ButtonGroup();
+		installPanel.removeAll();
+		for( LibraryVersion version : availableVersions) {
+			JRadioButton btn = new JRadioButton(version.toString());
+			btn.setToolTipText(version.getNote());
+			btn.setSelected(version.active);
+			versionGroup.add(btn);
+			buttons.add(btn);
+			btn.addActionListener(e -> {
+				if(btn.isSelected()) {
+					activateVersion(version);
+				}
+			});
+		}
+		updateFilter();
+	}
+
+	private void activateVersion(LibraryVersion version) {
+		if(version.active) {
+			System.out.println("[WARNING] Cannot activate version, already active: " + version);
+			return;
+		}
+		if(!version.downloaded) {
+			try {
+				tensorFlowInstallationService.downloadLib(new URL(version.url));
```

Comment:
Should happen in a separate thread (this freezes the UI).

Reply:
Thanks, done :)